File: Leetcode0174.java
package com.dezhonger.y2020.m02.d28;

/**
 * Created by dezhonger on 2020/02/28
 *
 * @author dezhonger
 * @since 2020/02/28
 */
public class Leetcode0174 {

    public int calculateMinimumHP(int[][] dungeon) {
        int m = dungeon.length;
        int n = dungeon[0].length;
        int[][] dp = new int[m ][n];
        dp[m - 1][n - 1] = Math.max(1, 1 - dungeon[m - 1][n - 1]);
        for (int i = m - 1; i >= 0; i--) {
            for (int j = n - 1; j >= 0; j--) {
                if (i == m - 1 && j == n - 1 ) continue;
                dp[i][j] = Integer.MAX_VALUE;
                if (i < m - 1) dp[i][j] = Math.min(dp[i][j], dp[i + 1][j] - dungeon[i][j]);
                if (j < n - 1) dp[i][j] = Math.min(dp[i][j], dp[i][j + 1] - dungeon[i][j]);
                dp[i][j] = Math.max(1, dp[i][j]);
            }
        }
        return dp[0][0];
    }
}

File: Leetcode0174.cpp
#include <bits/stdc++.h>
using namespace std;
//Filename: Leetcode0174.cpp
//
//Author: dezhonger - csdisassembly@gmail.com
//Create: 2021-05-14 21:00:30
//Last modified: 2021-05-14 21:10:38

class Solution {
public:
    int calculateMinimumHP(vector<vector<int>>& dungeon) {
        int n = dungeon.size(), m = dungeon[0].size();
        vector<vector<int>> f(n, vector<int>(m, 1e8));
        //f[i][j] 从[i, j]达到终点需要最少的血量值
        int c = dungeon[n - 1][m - 1];
        f[n - 1][m - 1] = c > 0 ? 1 : 1 - c;
        for (int i = n - 1; i >= 0; i--) {
            for (int j = m - 1; j >= 0; j--) {
                if (i + 1 < n) {
                    //f[i][j] + d[i][j] >= f[i + 1][j]
                    f[i][j] = min(f[i][j], max(1, f[i + 1][j] - dungeon[i][j]));
                }
                if (j + 1 < m) {
                    f[i][j] = min(f[i][j], max(1, f[i][j + 1] - dungeon[i][j]));
                }
            }
        }
        return f[0][0];
    }
};


